### ISSUE TITLE
Adapt plugins to new Plugin API introduced by #3210

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#3813]
Issue #3210  and PR ~~#3812~~ #3895 change the Plugin API to:

1. add `stop` (public), `stop?` (public) to LogStash::Inputs::Base
2. rename `teardown` to `close`
3. remove `shutdown`, `finished`. `finished?`, `running?`, `terminating?` from LogStash::Plugin

Refactor work needed on the plugins
---------------

#### [Input plugins](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+user%3Alogstash-plugins++label%3A%22shutdown+semantics%22)

Input plugins are now shutdown by an external call to `plugin.stop` instead of catching LogStash::ShutdownSignal exception.
Unless overridden, `stop` will simply makes `stop?` return `true`, thus allowing `run` to poll this and return after seeing the change.

In some plugins extra work must be done in `stop` to instruct `run` that it's time to return. For example, in the `logstash-input-udp` it's necessary to call `@socket.close` to make the blocking read on the socket raise an exception, thus breaking out the loop.
So, different input plugins will require different `stop` strategies.

Refactoring an input plugin involves:
* removing rescue of `ShutdownSignal` exception
* understand the nature of the `run` loop: is it a `while`? a `consumer.subscribe {|event| }`? is there a blocking operation on a socket/fd?
* use  `stop?` and/or override `stop` to make run return
* put any other cleanup/bookkeeping tasks in `close` (currently done in `teardown`)
* remove any calls to `shutdown`, `finished`, `finished?`, `running?` or `terminating?`

Then for testing you can use the shared example provided in https://github.com/elastic/logstash-devutils/pull/32 in the following manner:

```ruby
describe LogStash::Inputs::Http do
  let(:port) { rand(5000) + 1025 }
  it_behaves_like "an interruptible input plugin" do
    let(:config) { { "port" => port } }
  end
end
```

##### Testing the refactor

1) clone logstash, switch to PR 3895
2) clone logstash-devutils, switch to PR 32
3) clone plugin repository, switch to refactor PR (check above)
4) edit Gemfile to look like:
```
source 'https://rubygems.org'
gemspec
gem "logstash-core", :path => "/path/to/git/project/logstash"
gem "logstash-devutils", :path => "/path/to/git/project/logstash-devutils"
```
5) run `bundle install`
6) run `bundle exec rspec`

#### Filters and Outputs

Both will still shutdown using the ShutdownEvent sent by the pipeline so no major changes necessary.
The work in these plugins is:
* to remove calls that are no longer in the pipeline <-> plugin contract: `shutdown`, `finished`, `finished?`, `running?` or `terminating?`
* rename `teardown` to `close`

-------------
##### 1.1 Default Input Plugins

- [x] [logstash-input-couchdb_changes](https://github.com/logstash-plugins/logstash-input-couchdb_changes/issues/20) PR: https://github.com/logstash-plugins/logstash-input-couchdb_changes/pull/21
- [x] [logstash-input-elasticsearch](https://github.com/logstash-plugins/logstash-input-elasticsearch/issues/29) PR: https://github.com/logstash-plugins/logstash-input-elasticsearch/pull/28
- [x] [logstash-input-eventlog](https://github.com/logstash-plugins/logstash-input-eventlog/issues/18): PR https://github.com/logstash-plugins/logstash-input-eventlog/pull/19
- [x] [logstash-input-exec](https://github.com/logstash-plugins/logstash-input-exec/issues/3): PR https://github.com/logstash-plugins/logstash-input-exec/pull/2
- [x] [logstash-input-file](https://github.com/logstash-plugins/logstash-input-file/issues/68): PR https://github.com/logstash-plugins/logstash-input-file/pull/67
- [x] [logstash-input-ganglia](https://github.com/logstash-plugins/logstash-input-ganglia/issues/9) PR: https://github.com/logstash-plugins/logstash-input-ganglia/pull/10
- [x] [logstash-input-gelf](https://github.com/logstash-plugins/logstash-input-gelf/issues/17) PR: https://github.com/logstash-plugins/logstash-input-gelf/pull/19
- [x] [logstash-input-generator](https://github.com/logstash-plugins/logstash-input-generator/issues/10): PR https://github.com/logstash-plugins/logstash-input-generator/pull/9
- [x] [logstash-input-graphite](https://github.com/logstash-plugins/logstash-input-graphite/issues/4) 
- [x] [logstash-input-heartbeat](https://github.com/logstash-plugins/logstash-input-heartbeat/issues/8): PR https://github.com/logstash-plugins/logstash-input-heartbeat/pull/7
- [x] [logstash-input-http](https://github.com/logstash-plugins/logstash-input-http/issues/32): PR https://github.com/logstash-plugins/logstash-input-http/pull/31
- [x] [logstash-input-imap](https://github.com/logstash-plugins/logstash-input-imap/issues/10) PR:https://github.com/logstash-plugins/logstash-input-imap/pull/9
- [x] [logstash-input-irc](https://github.com/logstash-plugins/logstash-input-irc/issues/14) PR: https://github.com/logstash-plugins/logstash-input-irc/issues/14
- [x] [logstash-input-kafka](https://github.com/logstash-plugins/logstash-input-kafka/issues/41) PR: https://github.com/logstash-plugins/logstash-input-kafka/pull/42
- [x] [logstash-input-log4j](https://github.com/logstash-plugins/logstash-input-log4j/issues/19) PR:https://github.com/logstash-plugins/logstash-input-log4j/pull/17
- [x] [logstash-input-lumberjack](https://github.com/logstash-plugins/logstash-input-lumberjack/issues/52) PR: https://github.com/logstash-plugins/logstash-input-lumberjack/pull/59
- [x] [logstash-input-pipe](https://github.com/logstash-plugins/logstash-input-pipe/issues/11) PR: https://github.com/logstash-plugins/logstash-input-pipe/pull/10
- [x] [logstash-input-rabbitmq](https://github.com/logstash-plugins/logstash-input-rabbitmq/issues/38) PR: https://github.com/logstash-plugins/logstash-input-rabbitmq/pull/40
- [x] [logstash-input-redis](https://github.com/logstash-plugins/logstash-input-redis/issues/26) PR: https://github.com/logstash-plugins/logstash-input-redis/pull/27
- [x] [logstash-input-s3](https://github.com/logstash-plugins/logstash-input-s3/issues/53) PR: https://github.com/logstash-plugins/logstash-input-s3/pull/55
- [x] [logstash-input-snmptrap](https://github.com/logstash-plugins/logstash-input-snmptrap/issues/11) PR:https://github.com/logstash-plugins/logstash-input-snmptrap/pull/12
- [x] [logstash-input-sqs](https://github.com/logstash-plugins/logstash-input-sqs/issues/19) PR: https://github.com/logstash-plugins/logstash-input-sqs/pull/20
- [x] [logstash-input-stdin](https://github.com/logstash-plugins/logstash-input-stdin/issues/5): PR https://github.com/logstash-plugins/logstash-input-stdin/pull/4
- [x] [logstash-input-syslog](https://github.com/logstash-plugins/logstash-input-syslog/issues/20) PR: https://github.com/logstash-plugins/logstash-input-syslog/pull/21
- [x] [logstash-input-tcp](https://github.com/logstash-plugins/logstash-input-tcp/issues/17) PR:https://github.com/logstash-plugins/logstash-input-tcp/pull/8
- [x] [logstash-input-twitter](https://github.com/logstash-plugins/logstash-input-twitter/issues/27) PR: https://github.com/logstash-plugins/logstash-input-twitter/pull/28
- [x] [logstash-input-udp](https://github.com/logstash-plugins/logstash-input-udp/issues/8): PR https://github.com/logstash-plugins/logstash-input-udp/pull/7
- [x] [logstash-input-unix](https://github.com/logstash-plugins/logstash-input-unix/issues/9) PR: https://github.com/logstash-plugins/logstash-input-unix/pull/10
- [x] [logstash-input-xmpp](https://github.com/logstash-plugins/logstash-input-xmpp/issues/3) PR: https://github.com/logstash-plugins/logstash-input-xmpp/pull/4
- [x] [logstash-input-zeromq](https://github.com/logstash-plugins/logstash-input-zeromq/issues/9) PR:https://github.com/logstash-plugins/logstash-input-zeromq/pull/10

##### 1. Remove `shutdown`

Nothing to do.

##### 2. Remove `finished`

- [x] [logstash-output-file](https://github.com/logstash-plugins/logstash-output-file):  PLEASE REVIEW PR https://github.com/logstash-plugins/logstash-output-file/pull/16
- [x] [logstash-output-kafka](https://github.com/logstash-plugins/logstash-output-kafka): PLEASE REVIEW PR https://github.com/logstash-plugins/logstash-output-kafka/pull/28
- [x] [logstash-output-lumberjack](https://github.com/logstash-plugins/logstash-output-lumberjack): PLEASE REVIEW PR https://github.com/logstash-plugins/logstash-output-lumberjack/pull/10
- [x] [logstash-output-pipe](https://github.com/logstash-plugins/logstash-output-pipe):    PLEASE REVIEW https://github.com/logstash-plugins/logstash-output-pipe/pull/3
- [x] [logstash-output-rabbitmq](https://github.com/logstash-plugins/logstash-output-rabbitmq):    Please Review https://github.com/logstash-plugins/logstash-output-rabbitmq/pull/23
- [x] [logstash-output-s3](https://github.com/logstash-plugins/logstash-output-s3):    PLEASE REVIEW https://github.com/logstash-plugins/logstash-output-s3/pull/30
- [x] [logstash-output-sqs](https://github.com/logstash-plugins/logstash-output-sqs):    PLEASE REVIEW https://github.com/logstash-plugins/logstash-output-sqs/pull/8
- [x] [logstash-output-stdout](https://github.com/logstash-plugins/logstash-output-stdout): PLEASE REVIEW PR https://github.com/logstash-plugins/logstash-output-stdout/pull/4
- [x] [logstash-output-udp](https://github.com/logstash-plugins/logstash-output-udp):      PLEASE REVIEW PR https://github.com/logstash-plugins/logstash-output-udp/pull/4

##### 3. Remove `finished?`

Nothing to do.

##### 4. Remove `running?`

Nothing to do.

##### 5. Remove `terminating?`

Nothing to do.

##### 6. Rename `teardown` to `close`

Issue https://github.com/elastic/logstash/issues/3952 tracks this

#### 7. Other tasks

- [x] Improve the [logstash-input-example](https://github.com/logstash-plugins/logstash-input-example) documentation to reflect the new API
- [ ] Add documentation in elastic.co reference about new API signatures
- [ ] Blog about API changes
- [ ] Fix plugins not included by default in logstash releases - https://github.com/elastic/logstash/issues/3963

**Comment by ph** [comment]
Awesome @jsvd!

**Comment by suyograo** [comment]
@jsvd thanks for the detailed analysis! I added 2 more points at the end

**Comment by jsvd** [comment]
For all interested in testing this:

1) clone logstash, switch to PR 3812
2) clone logstash-devutils, switch to PR 32
3) clone plugin repository, switch to refactor PR (check above)
4) edit Gemfile to look like:
```
source 'https://rubygems.org'
gemspec
gem "logstash-core", :path => "/path/to/git/project/logstash"
gem "logstash-devutils", :path => "/path/to/git/project/logstash-devutils"
```
5) run `bundle install`
6) run `bundle exec rspec`

**Comment by purbon** [comment]
@jsvd work on a few inputs here, waiting to get more feedback from you before moving forward with more. Open question for me here is how deep to go on the "cleanup" :smile_cat:

**Comment by colinsurprenant** [comment]
before moving forward too deep (in particular for `close`/`teardown` and `stop?` we need to settle on #3810 and #3812 

@purbon we should avoid mixing cleanup & refactor work. one or the other but not both in the same PR, start with refactor since this is what is required here. after we can add/merge cleanups.

**Comment by colinsurprenant** [comment]
please note that some **input** plugins might need special handling/testing for when in blocked IO state to see if they can be unblocked from the pipeline thread form the `stop` method.

**Comment by jsvd** [comment]
@colinsurprenant indeed. the one exception I'm not worried about is the logstash-input-stdin, since it's equally non interruptible with both the new and old apis. Even with the PR you created to use the java interopt to extract the channel and make it interruptible, it's a non breaking change so it doesn't need to be rushed, IMO.

**Comment by colinsurprenant** [comment]
@jsvd +1

**Comment by wiibaa** [comment]
@colinsurprenant sorry to bump in, but as you are planning this huge refactoring, could you consider adding the small refactor described in #2447 about abstracting into base how to send an event to the filterworker(s) instead of exposing the queue in the run method. IMHO this could be the baseline for future enhancement on how inputs would be submitting events to the pipeline and allow to centralize ideas about detection/handling of back-pressure, batching of event, prioritization, dispatching, queue persistence etc etc. Thanks for reading!

**Comment by colinsurprenant** [comment]
@wiibaa thanks for the heads up - I will followup on #2447.

**Comment by colinsurprenant** [comment]
@jsvd @ph @jordansissel could use a review(s) of jordansissel/ruby-stud#27 to move forward with all plugins using `Stud.interval`

**Comment by andrewvc** [comment]
Is there a reason that 'stop' and 'close' have different names? I think this will be very confusing to future plugin authors. It's already confusing to me now. Ideally they would both also make `stop?` work correctly (some outputs sleep).

**Comment by talevy** [comment]
@andrewvc 

I think `teardown`(`close`) was originally intended for cleaning up, not instructing a plugin to effectively "stop" execution. one can be viewed as an "instruction" given to a running plugin. while `close` is an "instruction" given on a plugin that has been "stopped" (but might actually be running in the background via threads and open file handles).

I think it would be ok to rename `stop` to `close` for inputs. The only gotcha is that inputs that wish to override this and implement extra teardown actions will have to remember to call `super`. To avoid this, we can have the pipeline call another function that is not a part of the public api that will run the `@stop_called.make_true` stuff.

what do you think?

**Comment by andrewvc** [comment]
@talevy I'm not clear on why they need to call super? Presuming that they don't ever check `stop?` what would the issue be?

**Comment by talevy** [comment]
I don't like the fact that people who wish to use `#stop` as a place to teardown have to know about the guts of state management around `@stop_called`. It is true, if `stop?` is never used, this is a non-issue. I just don't think it is very nice to have to know whether this is a non-issue or an issue.

**Comment by andrewvc** [comment]
@talevy I misread your original comment. I do think we should have another function called externally that makes `stop?` return true. I think we should have a `__step_stop?` or the like function that does this that no one ever overrides.

**Comment by talevy** [comment]
disagree on the naming :) but yeah, that is what I meant!

**Comment by andrewvc** [comment]
@talevy I meant `__set_stop(true)`.

**Comment by jsvd** [comment]
all default plugins now have been refactored to address the new api. tracking non-default plugins in https://github.com/elastic/logstash/issues/3963
